sql/parser/eval.go

```diff
@@ -322,167 +322,152 @@ var cmpOpResultType = reflect.New(reflect.TypeOf(cmpOp{}).Field(0).Type.Out(0)).
 			return DBool(left.(DInterval).Duration < right.(DInterval).Duration), nil
 		},
 	},
 
 	cmpArgs{LE, stringType, stringType}: {
-		fn: func(left Datum, right Datum, _ *interface{}) (DBool, error) {
+		fn: func(_ EvalContext, left Datum, right Datum) (DBool, error) {
 			return DBool(left.(DString) <= right.(DString)), nil
 		},
 	},
 	cmpArgs{LE, bytesType, bytesType}: {
-		fn: func(left Datum, right Datum, _ *interface{}) (DBool, error) {
+		fn: func(_ EvalContext, left Datum, right Datum) (DBool, error) {
 			return DBool(left.(DBytes) <= right.(DBytes)), nil
 		},
 	},
 	cmpArgs{LE, boolType, boolType}: {
-		fn: func(left Datum, right Datum, _ *interface{}) (DBool, error) {
+		fn: func(_ EvalContext, left Datum, right Datum) (DBool, error) {
 			return DBool(!left.(DBool) || right.(DBool)), nil
 		},
 	},
 	cmpArgs{LE, intType, intType}: {
-		fn: func(left Datum, right Datum, _ *interface{}) (DBool, error) {
+		fn: func(_ EvalContext, left Datum, right Datum) (DBool, error) {
 			return DBool(left.(DInt) <= right.(DInt)), nil
 		},
 	},
 	cmpArgs{LE, floatType, floatType}: {
-		fn: func(left Datum, right Datum, _ *interface{}) (DBool, error) {
+		fn: func(_ EvalContext, left Datum, right Datum) (DBool, error) {
 			return DBool(left.(DFloat) <= right.(DFloat)), nil
 		},
 	},
 	cmpArgs{LE, dateType, dateType}: {
-		fn: func(left Datum, right Datum, _ *interface{}) (DBool, error) {
+		fn: func(_ EvalContext, left Datum, right Datum) (DBool, error) {
 			return DBool(left.(DDate) <= right.(DDate)), nil
 		},
 	},
 	cmpArgs{LE, timestampType, timestampType}: {
-		fn: func(left Datum, right Datum, _ *interface{}) (DBool, error) {
+		fn: func(_ EvalContext, left Datum, right Datum) (DBool, error) {
 			return !DBool(right.(DTimestamp).Before(left.(DTimestamp).Time)), nil
 		},
 	},
 	cmpArgs{LE, intervalType, intervalType}: {
-		fn: func(left Datum, right Datum, _ *interface{}) (DBool, error) {
+		fn: func(_ EvalContext, left Datum, right Datum) (DBool, error) {
 			return DBool(left.(DInterval).Duration <= right.(DInterval).Duration), nil
 		},
 	},
 
 	cmpArgs{Like, stringType, stringType}: {
-		fn: func(left Datum, right Datum, cache *interface{}) (DBool, error) {
-			var re *regexp.Regexp
-			switch t := (*cache).(type) {
-			case *regexp.Regexp:
-				re = t
-			default:
-				pattern := regexp.QuoteMeta(string(right.(DString)))
-				// Replace LIKE specific wildcards with standard wildcards
-				pattern = strings.Replace(pattern, "%", ".*", -1)
-				pattern = strings.Replace(pattern, "_", ".", -1)
-				pattern = anchorPattern(pattern, false)
-
-				re = regexp.MustCompile(pattern)
-				*cache = re
+		fn: func(ctx EvalContext, left Datum, right Datum) (DBool, error) {
+			pattern := regexp.QuoteMeta(string(right.(DString)))
+			// Replace LIKE specific wildcards with standard wildcards
+			pattern = strings.Replace(pattern, "%", ".*", -1)
+			pattern = strings.Replace(pattern, "_", ".", -1)
+			pattern = anchorPattern(pattern, false)
```

Comment:
Regenerating the regex pattern from the raw pattern on every invocation is unfortunate. Perhaps `GetRegexp` should take some sort of interface (perhaps even `interface{}`) instead of a `string`.